docs(36/WAKU2-BINDINGS-API): add filter, store, and lightpush functions (#525)
vacp2p/rfc: content/docs/rfcs/36/README.md

- `data`: Decrypted message payload base64 encoded,
- `padding`: Padding base64 encoded.

### `FilterSubscription` type

The criteria to create subscription to a light node in JSON Format:

```ts
{
    contentFilters: ContentFilter[];

>     topic: string?;

Reviewer: ```suggestion
    pubsubTopic: string?;
```

As we have two types of topics (pubsub, content), I think it makes it easier to new developers to always qualify the topic.

}
```

Fields:

- `contentFilters`: Array of [`ContentFilter`](#contentfilter-type) being subscribed to / unsubscribed from.

> - `topic`: Optional message topic.

Reviewer: ```suggestion
- `topic`: Optional pubsub topic.
```

### `ContentFilter` type

```ts
{
    contentTopic: string;
}
```

Fields:

> - `contentTopic`: The content topic of a waku message.

Reviewer: ```suggestion
- `contentTopic`: The content topic of a Waku message.
```

### `StoreQuery` type

> Criteria used to retreive historical messages

Reviewer 2: ```suggestion
Criteria used to retrieve historical messages
```

```ts
interface StoreQuery {
    pubsubTopic?: string;
    contentFilters?: ContentFilter[];

>     startTime?: number;
>     endTime?: number;

Reviewer: ```suggestion
   timeFilter?: {
       startTime: number;
       endTime: number;
   }
```

You have to set both or neither. To avoid user error I recommend modifying the API so that the user cannot make a mistake without breaking typing.

Author: According to this code, it should be possible to set either:
https://github.com/status-im/nwaku/commit/2cb7123df84b18ab7e4df0be0d77353cca4820f2#diff-9adb9e1f6697143d9e2f5e5870293666b221fe21047b4ca05e426130329068c7R223-R231

Reviewer: > According to this code, it should be possible to set either: [status-im/nwaku@2cb7123#diff-9adb9e1f6697143d9e2f5e5870293666b221fe21047b4ca05e426130329068c7R223-R231](https://github.com/status-im/nwaku/commit/2cb7123df84b18ab7e4df0be0d77353cca4820f2#diff-9adb9e1f6697143d9e2f5e5870293666b221fe21047b4ca05e426130329068c7R223-R231)

Well, this code needs to be fixed :)

From https://rfc.vac.dev/spec/21/
> A time-based query is considered valid if its end_time is larger than or equal to the start_time. Queries that do not adhere to this condition will not get through **e.g. an open-end time query in which the start_time is given but no end_time is supplied is not valid.** If both start_time and end_time are omitted then no time-window filter takes place.*

Cc @LNSD

- `endTime`: The inclusive upper bound on the timestamp of queried messages. This field holds the Unix epoch time in nanoseconds.
- `pagingOptions`: Paging information in [`PagingOptions`](#pagingoptions-type) format.

### `StoreResponse` type

The response received after doing a query to a store node:

```ts
interface StoreResponse {
    messages: JsonMessage[];
    pagingOptions?: PagingOptions;
  }
```
Fields:

- `messages`: Array of retrieved historical messages in [`JsonMessage`](#jsonmessage-type) format.
- `pagingOption`: Paging information in [`PagingOptions`](#pagingoptions-type) format from which to resume further historical queries

### `PagingOptions` type

```ts
interface PagingOptions {
    pageSize: number;
    cursor?: Index;
    forward: bool;
  }
```
Fields:

- `pageSize`: Number of messages to retrieve per page.
- `cursor`: Message Index from which to perform pagination. If not included and forward is set to true, paging will be performed from the beginning of the list. If not included and forward is set to false, paging will be performed from the end of the list.

> - `forward`: 	`true` if paging forward, `false` if paging backward

Reviewer: ```suggestion
- `forward`:  `true` if paging forward, `false` if paging backward
```

### `Index` type

```ts
interface Index {
    digest: string;
    receiverTime: number;
    senderTime: number;
    pubsubTopic: string;
  }
```

Fields:

- `digest`: Hash of the message at this [`Index`](#index-type).

> - `receiverTime`: UNIX timestamp in nanoseconds at which the message at this [`Index`](#index-type) was received

Reviewer: ```suggestion
- `receiverTime`: UNIX timestamp in nanoseconds at which the message at this [`Index`](#index-type) was received.
```

```

## Waku Filter

> ### `extern char* waku_filter_subscribe(char* filterJSON, char* peerID, int ms)`

Reviewer: ```suggestion
### `extern char* waku_filter_subscribe(char* filterJSON, char* peerID, int timeoutMs)`
```

Same var name than other APIs.

Creates a subscription in a lightnode for messages that matches a content filter and optionally a [PubSub `topic`](https://github.com/libp2p/specs/blob/master/pubsub/README.md#the-topic-descriptor).

**Parameters**

> 1. `char* filterJSON`: JSON string containing the [`FilterSubscription`](#filtersubscription-type) to unsubscribe from

Reviewer: ```suggestion
1. `char* filterJSON`: JSON string containing the [`FilterSubscription`](#filtersubscription-type) to subscribe to.
```

   Use `NULL` to automatically select a node.
3. `int timeoutMs`: Timeout value in milliseconds to execute the call.
   If the function execution takes longer than this value,
   the execution will be canceled and an error returned.
   Use `0` for no timeout.

**Returns**

A [`JsonResponse`](#jsonresponse-type).
If the execution is successful, the `result` field is set to `true`.

For example:

```json
{
  "result": true
}
```

**Events**

When a message is received, a ``"message"` event` is emitted containing the message, pubsub topic, and node ID in which
the message was received.

The `event` type is [`JsonMessageEvent`](#jsonmessageevent-type).

For Example:

```json
{
  "type": "message",
  "event": {
    "subscriptionID": 1,
    "pubsubTopic": "/waku/2/default-waku/proto",
    "messageID": "0x6496491e40dbe0b6c3a2198c2426b16301688a2daebc4f57ad7706115eac3ad1",
    "wakuMessage": {
      "payload": "TODO",
      "contentTopic": "/my-app/1/notification/proto",
      "version": 1,
      "timestamp": 1647826358000000000
    }
  }
}
```

### `extern char* waku_filter_unsubscribe(char* filterJSON, int timeoutMs)`

> Removes subscriptions in a light node matching a content filter and, optionally, a [PubSub `topic`](https://github.com/libp2p/specs/blob/master/pubsub/README.md#the-topic-descriptor).

Reviewer 2: Is the idea that it unsubscribes from all peers where a matching filter exists? Asking because `subscribe` can be peer-specific, but `unsubscribe` is not.

Author: Yes, it would try to remove from all peers where such subscription exists. I tried to make the API similar to [16/WAKU2-RPC
](https://rfc.vac.dev/spec/16/#delete_waku_v2_filter_v1_subscription)

Reviewer 2: Makes sense. Thanks!

Optionally sign,
encrypt using asymmetric encryption
and publish a message using Waku Lightpush.

**Parameters**

1. `char* messageJson`: JSON string containing the [Waku Message](https://rfc.vac.dev/spec/14/) as [`JsonMessage`](#jsonmessage-type).
2. `char* pubsubTopic`: pubsub topic on which to publish the message.
   If `NULL`, it uses the default pubsub topic.
3. `char* peerID`: Peer ID supporting the lightpush protocol.
   The peer must be already known.
   It must have been added before with [`waku_add_peer`](#extern-char-waku_add_peerchar-address-char-protocolid)
   or previously dialed with [`waku_connect_peer`](#extern-char-waku_connect_peerchar-address-int-timeoutms).
4. `char* publicKey`: hex encoded public key to be used for encryption.
5. `char* optionalSigningKey`: hex encoded private key to be used to sign the message.
6. `int timeoutMs`: Timeout value in milliseconds to execute the call.
   If the function execution takes longer than this value,
   the execution will be canceled and an error returned.
   Use `0` for no timeout.

Note: `messageJson.version` is overwritten to `1`.

**Returns**

A [`JsonResponse`](#jsonresponse-type).

> If the execution is successful, the `result` field contains the message ID.

Reviewer: What is a message ID? is it a string or number?

Author: A hex string

Reviewer: Can we have a sample so that it's clear that a hex string is expected?
